src/react/utils.js

```diff
@@ -58,6 +60,92 @@ export function valueIsClassComponent(realm: Realm, value: Value) {
   return false;
 }
 
+function canHoistObject(realm: Realm, object: ObjectValue, hoistedScope?: Generator): boolean {
+  if (isReactElement(object)) {
+    return canHoistReactElement(realm, object);
+  }
+  for (let [propName] of object.properties) {
+    let prop = Get(realm, object, propName);
+    if (prop instanceof AbstractValue) {
+      return false;
+    } else if (prop instanceof FunctionValue && prop.parent !== hoistedScope) {
+      return false;
+    } else if (prop instanceof ArrayValue) {
+      let isHoistable = canHoistArray(realm, prop, hoistedScope);
+
+      if (!isHoistable) {
+        return false;
+      }
+    } else if (prop instanceof ObjectValue) {
+      let isHoistable = canHoistObject(realm, prop, hoistedScope);
+
+      if (!isHoistable) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+function canHoistArray(realm: Realm, array: ArrayValue, hoistedScope?: Generator): boolean {
+  let lengthValue = Get(realm, array, "length");
+  invariant(lengthValue instanceof NumberValue);
+  let length = lengthValue.value;
+  for (let i = 0; i < length; i++) {
+    let element = Get(realm, array, "" + i);
+
+    if (element instanceof AbstractValue) {
```

Comment:
This seems like it could be factored out into a helper function.

Reply:
I thought about it, but it actually makes the logic workflow harder to read in my opinion.